### PR TITLE
fix: saving & restoring RNG state for Keras & Estimators [DET-3743]

### PR DESCRIPTION
**Pull request by mackrorysd**

### CODE REVIEW

File: harness/determined/estimator/_estimator_trial.py [hunk 1]
```diff
@@ -534,6 +576,8 @@ def _init_model(self) -> None:
         self._init_val_hooks()
         self._init_paths()
 
+        self.train_hooks.append(RNGStateHook(self.estimator_dir))
```
Comment (shiyuann) [inline harness/determined/estimator/_estimator_trial.py:579]: blocking: this probably should be embedded into `DeterminedControlHook` rather than stand-alone.

File: e2e_tests/tests/fixtures/keras_no_op/model_def.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        8,
+                        8,
+                    ),
+                )
+            ]
+        )
+        model = self.context.wrap_model(model)
+        model.compile(
+            loss=tf.keras.losses.MeanSquaredError(),
+            metrics=[
```
Comment (shiyuann) [inline e2e_tests/tests/fixtures/keras_no_op/model_def.py:65]: q: are the `result()` of those metrics called once per batch?

Comment (mackrorysd) [reply e2e_tests/tests/fixtures/keras_no_op/model_def.py:65]: Yes, I believe so.

File: harness/determined/estimator/_estimator_trial.py [hunk 3]
```diff
@@ -23,6 +24,47 @@
...
+    def __init__(self, checkpoint_dir: pathlib.Path):
+        try:
+            self.rng_state = None
+            with open(checkpoint_dir.joinpath("rng_state.pkl"), "rb") as f:
+                self.rng_state = pickle.load(f)
+        except IOError:
+            logging.debug("No RNG state found in checkpoint_dir")
+
+    def after_create_session(self, session, coord):  # type: ignore
+        if self.rng_state is not None:
```
[the remark this answers is not in the document]
Comment (mackrorysd) [reply harness/determined/estimator/_estimator_trial.py:37]: There won't always be an RNG state, because it might be an old checkpoint, and of course after_create_session is also called on new experiments. So I was using the existence of rng_state as the signal for that. If it doesn't exist, I already log a warning. Is there a better signal to use?

File: harness/determined/keras/_tf_keras_trial.py [hunk 4]
```diff
@@ -401,6 +402,23 @@ def _load(self) -> None:
...
+
+            if version.parse(tf.__version__) < version.parse("2.0.0"):
+                tf.random.set_random_seed(rng_state["tf_rng_global_seed"])
+            else:
+                algorithm = rng_state["tf2_rng_global_algorithm"]
+                state = rng_state["tf2_rng_global_state"]
+                generator = tf.random.Generator.from_state(state, algorithm)
+                tf.random.set_global_generator(generator)
+        except IOError:
+            logging.warn("Checkpoint did not include RNG state")
```
[the remark this answers is not in the document]
Comment (mackrorysd) [reply harness/determined/keras/_tf_keras_trial.py:420]: Yeah see comment below - there are times when no rng state is valid.

File: harness/determined/estimator/_estimator_trial.py (hunk 1 above)
Comment (mackrorysd) [reply harness/determined/estimator/_estimator_trial.py]: Yeah that makes sense. I had to shuffle a couple of things because it depends on parts of init_paths() which depends on ini_train_hooks. Some harness tests failed locally but I'm not certain they're related - I've had some flakiness running them on my machine. Let's see how Circle does...

File: harness/determined/estimator/_estimator_trial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/estimator/_estimator_trial.py]: I was thinking `self.rng_state` should be initialized when creating an estimator trial controller but I saw in the latest version you have removed this internal state so it looks good to me.

File: harness/determined/estimator/_estimator_trial.py [hunk 5]
```diff
@@ -293,6 +294,48 @@ def control_loop(self) -> None:
...
+
+    def on_checkpoint_end(self, checkpoint_dir: str) -> None:
+        self.save_rng_state_with_checkpoint(checkpoint_dir)
+
+    def load_rng_state_from_checkpoint(self, checkpoint_dir: str) -> None:
+        rng_state = None
+        try:
+            with open(checkpoint_dir + "/rng_state.pkl", "rb") as f:
+                rng_state = pickle.load(f)
+        except IOError:
```
Comment (shiyuann) [inline harness/determined/estimator/_estimator_trial.py:308]: can we add a comment here to say this is only for backwards compatibility? because I thought this logic should be removed sometime in the future

File: e2e_tests/tests/experiment/test_noop.py [hunk 6]
```diff
@@ -252,36 +252,42 @@ def test_startup_hook() -> None:
...
 
 
-@pytest.mark.e2e_cpu  # type: ignore
-def test_rng_restore() -> None:
+def _test_rng_restore(fixture: str, metrics: list) -> None:
+    """
+    This test confirms that an experiment can be restarted from a checkpoint
+    with the same RNG state. It requires a test fixture that will emit
+    random numbers from all of the RNGs used in the relevant framework as
+    metrics. The experiment must have a const.yaml, run for at least 3 steps,
```
Comment (shiyuann) [inline e2e_tests/tests/experiment/test_noop.py:260]: q: does this use validation metrics?

**Comment by shiyuann** [comment]
Just a couple of questions. Otherwise, it looks good!

File: e2e_tests/tests/experiment/test_noop.py (hunk 6 above)
Comment (mackrorysd) [reply e2e_tests/tests/experiment/test_noop.py:260]: No, I think they're all training metrics

File: harness/determined/estimator/_estimator_trial.py (hunk 5 above)
Comment (mackrorysd) [reply harness/determined/estimator/_estimator_trial.py:308]: Sure. Initially it was also to avoid problems on the first step of a new experiment, but now that it's in on_checkpoint_load() it is strictly for backwards compatibility.

I actually think this is much cleaner overall now that I'm using those 2 hooks.

Comment (mackrorysd) [reply harness/determined/estimator/_estimator_trial.py:308]: I also upgraded the log line to a warning again, since it's not supposed to happen on new experiments, and users of old checkpoints may want to be aware of this. I verified this code path is not active when running new experiments.

Comment (shiyuann) [reply harness/determined/estimator/_estimator_trial.py:308]: thanks! can we add a prefix "backward compatibility: " in the comment as in other places?

**Review by shiyuann — APPROVED** [review]
Looks good!
